# HW1

Submitted by:

Parthiv Borgohain (pb25347)

## Q1.1 
For the function y = f(x) = w.x + b, where w=[2,1] and b=3, nd the partial derivates of y w.r.t the components of x (i.e.: dy/dx1 and dy/dx2) at x = [4,2] both on paper and using PyTorch.

In [3]:
import torch
from torch import nn
from torch.utils.data import TensorDataset
from torch.utils.data import DataLoader

x=torch.tensor([4,2], dtype=torch.float, requires_grad=True) 
w=torch.tensor([2,1],dtype=torch.float)
b=torch.tensor(3,dtype=torch.float)
y=torch.matmul(w.T,x) + b
y.backward()


In [4]:
#Now let us compute the gradient
x.grad

tensor([2., 1.])

Hence, the two gradients w.r.t x1 and x2 are **2 and 1** respectively

## 1.2

For a model y = f(x), the predicted and true values are as follows: 

y_true = [0,1,1,0] 

y_pred = [0.1,0.95,1.10,0.2] 

Find the mean squared error both on paper and using PyTorch. In PyTorch, solve it using an inbuilt function and also by defining your own squared error function.

In [5]:
#Let us try to find Mean Square Error using an inbuilt function in Pytorch
y_true=torch.tensor([0,1,1,0],dtype=torch.float)
y_pred=torch.tensor([0.1,0.95,1.10,0.2])
loss=torch.nn.MSELoss()
loss(y_true,y_pred)

tensor(0.0156)

In [6]:
#Let us define a custom MSE function
def mseLoss(y_true,y_pred):
    mseloss=torch.mean((y_true-y_pred)*(y_true-y_pred))
    return mseloss

In [7]:
loss=mseLoss(y_true,y_pred)
loss

tensor(0.0156)

So, the MSE loss is obtained by both methods is approximately **0.0156**

## 1.3

You are given the following dataset of an XOR function: Assume that we fit a linear function, y = w.x + b to this dataset with w = [2,1] and b=3

i) Find the mean squared error (loss) over this dataset for the above weights and bias.



In [8]:
x=torch.tensor([[0,0],[0,1],[1,0],[1,1]],dtype=torch.float)
y_true=torch.tensor([0,1,1,0],dtype=torch.float)
w=torch.tensor([2,1],dtype=torch.float,requires_grad=True)
b=torch.tensor(3.0,requires_grad=True)
y_pred=torch.sum(w*x,1)+b

In [9]:
def mse_loss(y_true,y_pred):
    mseloss=torch.mean((y_true-y_pred)*(y_true-y_pred))
    return mseloss
def model(X):
    return torch.matmul(X,w.T) + b

In [10]:
loss=mse_loss(y_true,y_pred)
loss

tensor(17.5000, grad_fn=<MeanBackward0>)

So, the MSE obtained is **17.5** approximately

ii) Which direction should the weights move in to decrease the loss by maximum amount? Find that direction both on paper and using PyTorch.



In [11]:
data = TensorDataset(x, y_true)
print(loss.backward())
print(w.grad)
print(b.grad)

None
tensor([5.0000, 4.5000])
tensor(8.)


Gradients w.r.t w1 and w2 are **5 and 4.5** 

Gradient w.r.t b is **8**

The direction vector is **[5,4.5,8]**

iii) For what values of w and b is the loss minimum? Solve on paper only.

Has been solved in the attached pdf file

### Extra credit
Can you solve for the optimum w and b using PyTorch? Hint: use gradient descent

In [12]:
batch_size = 4
train_loader = DataLoader(data, batch_size=batch_size, shuffle=True)

#Let us try running for 1000 epochs using gradient descent and check if we can reach 0  loss
epochs = 1000
for i in range(epochs):
    # Iterate through training dataloader
    for x,y in train_loader:
        # Generate Prediction
        preds = model(x)
        # Get the loss and perform backpropagation
        loss = mse_loss(y,preds)
        loss.backward()
        # Let's update the weights
        with torch.no_grad():
            w -= w.grad *0.1
            b -= b.grad * 0.1
            # Set the gradients to zero
            w.grad.zero_()
            b.grad.zero_()
    print(f"Epoch {i}/{epochs}: Loss: {loss}")

Epoch 0/1000: Loss: 17.5
Epoch 1/1000: Loss: 2.6050000190734863
Epoch 2/1000: Loss: 1.3877873420715332
Epoch 3/1000: Loss: 0.8183243274688721
Epoch 4/1000: Loss: 0.5500246286392212
Epoch 5/1000: Loss: 0.42193183302879333
Epoch 6/1000: Loss: 0.3592788577079773
Epoch 7/1000: Loss: 0.32731887698173523
Epoch 8/1000: Loss: 0.30988913774490356
Epoch 9/1000: Loss: 0.29945817589759827
Epoch 10/1000: Loss: 0.29250460863113403
Epoch 11/1000: Loss: 0.28737133741378784
Epoch 12/1000: Loss: 0.28326886892318726
Epoch 13/1000: Loss: 0.2798125743865967
Epoch 14/1000: Loss: 0.27680760622024536
Epoch 15/1000: Loss: 0.2741485834121704
Epoch 16/1000: Loss: 0.27177348732948303
Epoch 17/1000: Loss: 0.26964133977890015
Epoch 18/1000: Loss: 0.2677222788333893
Epoch 19/1000: Loss: 0.26599282026290894
Epoch 20/1000: Loss: 0.2644330561161041
Epoch 21/1000: Loss: 0.2630259096622467
Epoch 22/1000: Loss: 0.26175615191459656
Epoch 23/1000: Loss: 0.26061028242111206
Epoch 24/1000: Loss: 0.2595761716365814
Epoch 25/10

In [13]:
for x,y in train_loader:
    preds = model(x)
    print("Prediction is :n",preds)
    print("nActual targets is :n",y)
    break

Prediction is :n tensor([0.5000, 0.5000, 0.5000, 0.5000], grad_fn=<AddBackward0>)
nActual targets is :n tensor([1., 0., 0., 1.])


In [14]:
# Finding optimal weight values
w 

tensor([1.2834e-07, 1.3101e-07], requires_grad=True)

In [15]:
#Finding optimal value of bias
b

tensor(0.5000, requires_grad=True)

## 2.1

2.1 Given a 1-d input x=[1,-1,3,4,4] and a kernel=[1,1], find the 1-d convolution for stride=1 and stride=2. Solve both on paper and using pyTorch.


In [18]:
x1 = torch.tensor([1,-1,3,4,4], dtype=torch.float)
x1=x1.unsqueeze(0)
x1=x1.unsqueeze(0)
x1=x1.squeeze(0)
cnn1d = nn.Conv1d(in_channels=1, out_channels=1, kernel_size=2,bias=False)
cnn1d.weight.data=torch.tensor([1,1],dtype=torch.float,requires_grad=True)
cnn1d.weight.data=cnn1d.weight.data.unsqueeze(0).unsqueeze(0)
cnn1d.weight.data
out_1=cnn1d(x1)
print("So, for stride=1, the convolved output is:",out_1)

So, for stride=1, the convolved output is: tensor([[0., 2., 7., 8.]], grad_fn=<SqueezeBackward1>)


In [19]:
cnn1d = nn.Conv1d(in_channels=1, out_channels=1, kernel_size=2, stride=2, bias=False)
cnn1d.weight.data=torch.tensor([1,1],dtype=torch.float,requires_grad=True)
cnn1d.weight.data=cnn1d.weight.data.unsqueeze(0).unsqueeze(0)
out_2=cnn1d(x1)
print("So, for stride =2, the convolved output is:",out_2)

So, for stride =2, the convolved output is: tensor([[0., 7.]], grad_fn=<SqueezeBackward1>)


## 2.2

You’re given the following grayscale image (any matrix in 2 dimensions is a grayscale image):

i) You’re given 2 2*2 Filters (kernels): [[1,0],[0,1]] and [[0,1],[1,0]]. Find the output when the image is convolved using each Filter with stride=1. 

In [20]:
x3 = torch.tensor([[0.1,-0.6,0.4,0.8],[-0.4,0.3,0.9,0.2],[0.5,0.2,0.8,-0.7],[0.3,0.7,-0.4,0.1]])
#Flattening input
x3=x3.unsqueeze(0)
cnn2d_2=nn.Conv2d(in_channels=1,out_channels=2,kernel_size=2,bias=False)
cnn2d_2.weight.data=torch.tensor([[[1,0],[0,1]],
                                  
                                  [[0,1],[1,0]]   ],dtype=torch.float)
cnn2d_2.weight.data=cnn2d_2.weight.data.unsqueeze(1)

In [21]:
out=cnn2d_2(x3)

In [22]:
print("Convolved Output Image for both channels is:\n",out)

Convolved Output Image for both channels is:
 tensor([[[ 0.4000,  0.3000,  0.6000],
         [-0.2000,  1.1000,  0.2000],
         [ 1.2000, -0.2000,  0.9000]],

        [[-1.0000,  0.7000,  1.7000],
         [ 0.8000,  1.1000,  1.0000],
         [ 0.5000,  1.5000, -1.1000]]], grad_fn=<SqueezeBackward1>)


ii) What are the dimensions of each channel in (i)? What will be the dimensions of the output when stride=2?

In [23]:
cnn2d_2_std2=nn.Conv2d(in_channels=1,out_channels=1,kernel_size=2,stride=2,bias=False)
cnn2d_2_std2.weight.data=torch.tensor([[[1,0],[0,1]],
                                  
                                  [[0,1],[1,0]]   ],dtype=torch.float)
cnn2d_2_std2.weight.data=cnn2d_2_std2.weight.data.unsqueeze(1)
out=cnn2d_2_std2(x3)
print("Output for stride=2 is",out) 

Output for stride=2 is tensor([[[ 0.4000,  0.6000],
         [ 1.2000,  0.9000]],

        [[-1.0000,  1.7000],
         [ 0.5000, -1.1000]]], grad_fn=<SqueezeBackward1>)


So, dimension of each channel when stride=1 is 3x3 and when stride=2 is 2x2

### iii) The given image is a 2-d matrix. How can you convolve it so that the output channel has only one dimension?

Solution is given in attached PDF.

### iv) Perform a 2*2 max-pooling with stride=1 on the image both on paper and using PyTorch. What are the dimensions of the channel after max-pooling?

In [24]:
maxpool = nn.MaxPool2d(2, stride=1)

In [25]:
out=maxpool(x3)
out

tensor([[[0.3000, 0.9000, 0.9000],
         [0.5000, 0.9000, 0.9000],
         [0.7000, 0.8000, 0.8000]]])

In [26]:
print("Dimensions of channel after max pooling",out.shape)

Dimensions of channel after max pooling torch.Size([1, 3, 3])


### Extra-credit
If the size of the output channel after performing a 2*2 convolution (stride=1) and a 2*2 max-pooling (stride=1) is the same, why do you think we needed max-pooling when we could perform convolution to decrease the size
of the image. Think in terms of the advantages that max-pooling may oer over convolution. When do you think a max-pooling operation may not be advantageous (and in fact may hurt the network)?

Solution is given in the attached PDF File

## Question 3 :

All answers have been solved in the attached PDF File.